watchdog.8
.TH WATCHDOG 8 "August 2013"
.UC 4
.SH NAME
watchdog \- a software watchdog daemon
.SH SYNOPSIS
.B watchdog
.RB [ \-F | \-\-foreground ]
.RB [ \-f | \-\-force ]
.RB [ \-c " \fIfilename\fR|" \-\-config\-file " \fIfilename\fR]"
.RB [ \-v | \-\-verbose ]
.RB [ \-s | \-\-sync ]
.RB [ \-b | \-\-softboot ] 
.RB [ \-q | \-\-no\-action ]
.SH DESCRIPTION
The Linux kernel can reset the system if serious problems are detected.
This can be implemented via special watchdog hardware, or via a slightly
less reliable software-only watchdog inside the kernel. Either way, there
needs to be a daemon that tells the kernel the system is working fine. If the
daemon stops doing that, the system is reset.
.PP
.B watchdog 
is such a daemon. It opens
.IR /dev/watchdog , 
and keeps writing to it often enough to keep the kernel from resetting,
at least once per minute. Each write delays the reboot
time another minute. After a minute of inactivity the watchdog hardware will
cause the reset. In the case of the software watchdog the ability to 
reboot will depend on the state of the machines and interrupts.
.PP
The watchdog daemon can be stopped without causing a reboot if the device 
.I /dev/watchdog
is closed correctly, unless your kernel is compiled with the
.I CONFIG_WATCHDOG_NOWAYOUT
option enabled.
.SH TESTS
The watchdog daemon does several tests to check the system status:
.IP \(bu 3
Is the process table full?
.IP \(bu 3
Is there enough free memory?
.IP \(bu 3
Is there enough allocatable memory?
.IP \(bu 3
Are some files accessible?
.IP \(bu 3
Have some files changed within a given interval?
.IP \(bu 3
Is the average work load too high?
.IP \(bu 3
Has a file table overflow occurred?
.IP \(bu 3
Is a process still running? The process is specified by a pid file.
.IP \(bu 3
Do some IP addresses answer to ping?
.IP \(bu 3
Do network interfaces receive traffic?
.IP \(bu 3
Is the temperature too high? (Temperature data not always available.)
.IP \(bu 3
Execute a user defined command to do arbitrary tests.
.IP \(bu 3
Execute one or more test/repair commands found in /etc/watchdog.d.  These commands are called with the argument \fBtest\fP or \fBrepair\fP.
.PP
If any of these checks fail watchdog will cause a shutdown. Should any of
these tests except the user defined binary last longer than one minute the
machine will be rebooted, too.
.PP
.SH OPTIONS
Available command line options are the following:
.TP
.BR \-v ", " \-\-verbose
Set verbose mode. Only implemented if compiled with 
.I SYSLOG 
feature. This
mode will log each several infos in 
.I LOG_DAEMON 
with priority 
.IR LOG_INFO.
This is useful if you want to see exactly what happened until the watchdog rebooted
the system. Currently it logs the temperature (if available), the load
average, the change date of the files it checks and how often it went to sleep.
.TP
.BR \-s ", " \-\-sync
Try to synchronize the filesystem every time the process is awake. Note that
the system is rebooted if for any reason the synchronizing lasts longer
than a minute.
.TP
.BR \-b ", " \-\-softboot
Soft-boot the system if an error occurs during the main loop, e.g. if a 
given file is not accessible via the 
.BR stat (2)
call. Note that
this does not apply to the opening of 
.I /dev/watchdog 
and 
.IR /proc/loadavg ,
which are opened before the main loop starts.
.TP
.BR \-F ", " \-\-foreground
Run in foreground mode, useful for running under systemd (for example).
.TP
.BR \-f ", " \-\-force
Force the usage of the interval given or the maximal load average given 
in the config file. Without this option these values are sanity checked.
.TP
.BR \-c " \fIconfig-file\fR, " \-\-config\-file " \fIconfig-file"
Use
.I config-file
as the configuration file instead of the default 
.IR /etc/watchdog.conf .
.TP
.BR \-q ", " \-\-no\-action
Do not reboot or halt the machine. This is for testing purposes. All checks
are executed and the results are logged as usual, but no action is taken.
Also your hardware card or the kernel software watchdog driver is not
enabled. Temperature checking is also disabled since this triggers
the hardware watchdog on some cards.
.SH FUNCTION
After
.B watchdog 
starts, it puts itself into the background and then tries all checks
specified in its configuration file in turn. Between each two tests it will
write to the kernel device to prevent a reset. After finishing all tests
watchdog goes to sleep for some time. The kernel drivers expects a write to the
watchdog device every minute. Otherwise the system will be reset. 
.B watchdog 
will sleep for a configure interval that defaults to 1 second to make sure it
triggers the device early enough.
.PP
Under high system load 
.B watchdog 
might be swapped out of memory and may fail
to make it back in in time. Under these circumstances the Linux kernel will
reset the machine. To make sure you won't get unnecessary reboots make
sure you have the variable 
.I realtime
set to 
.I yes 
in the configuration file
.IR watchdog.conf . 
This adds real time support to 
.BR watchdog :
it will lock itself into memory and there should  be no problem even under the
highest of loads.
.PP
On system running out of memory the kernel will try to free enough memory by killing process. The
.B watchdog
daemon itself is exempted from this so-called out-of-memory killer. 
.PP
Also you can specify a maximal allowed load average. Once this load average
is reached the system is rebooted. You may specify maximal load averages for
1 minute, 5 minutes or 15 minutes. The default values is to disable this
test. Be careful not to set this parameter too low. To set a value less then
the predefined minimal value of 2, you have to use the 
.B -f 
option.
.PP
You can also specify a minimal amount of virtual memory you want to have
available as free. As soon as more virtual memory is used action is taken by
.BR watchdog . 
Note, however, that watchdog does not distinguish between
different types of memory usage. It just checks for free virtual memory.
.PP
If you have a watchdog card with temperature sensor you can specify 
the maximal allowed temperature. Once this temperature is reached the
system is halted. The default value is 120. There is no unit conversion so make
sure you use the same unit as your hardware. 
.B watchdog 
will issue warnings
once the temperature increases 90%, 95% and 98% of this temperature.
.PP
When using file mode 
.B watchdog 
will try to
.BR stat (2)
the given files. Errors returned
by stat will 
.B not
cause a reboot. For a reboot the stat call has to last at least one minute.
This may happen if the file is located on an NFS mounted filesystem. If your
system relies on an NFS mounted filesystem you might try this option.
However, in such a case the 
.I sync 
option may not work if the NFS server is
not answering.
.PP
.B watchdog
can read the pid from a pid file and 
see whether the process still exists. If not, action is taken
by 
.BR watchdog . 
So you can for instance restart the server from your
.IR repair-binary .
.PP
.B watchdog 
will try periodically to fork itself to see whether the process
table is full. This process will leave a zombie process until watchdog wakes
up again and catches it; this is harmless, don't worry about it.
.PP
In ping mode 
.B watchdog 
tries to ping the given IP addresses. These addresses do
not have to be a single machine. It is possible to ping to a broadcast
address instead to see if at least one machine in a subnet is still living.
.PP
.B Do not use this broadcast ping unless your MIS person a) knows about it and
.B b) has given you explicit permission to use it!
.PP
.B watchdog 
will send out three ping packages and wait up to <interval> seconds
for the reply with <interval> being the time it goes to sleep between two
times triggering the watchdog device. Thus a unreachable network will not
cause a hard reset but a soft reboot.
.PP
You can also test passively for an unreachable network by just monitoring
a given interface for traffic. If no traffic arrives the network is
considered unreachable causing a soft reboot or action from the 
repair binary.
.PP
.B watchdog
can run an external command for user-defined tests. A return code not equal 0
means an error occurred and watchdog should react. If the external command is
killed by an uncaught signal this is considered an error by watchdog too.
The command may take longer than the time slice defined for the kernel device
without a problem. However, error messages are
generated into the syslog facility. If you have enabled softboot on error
the machine will be rebooted if the binary doesn't exit in half the time
.B watchdog 
sleeps between two tries triggering the kernel device.
.PP
If you specify a repair binary it will be started instead of shutting down
the system. If this binary is not able to fix the problem 
.B watchdog 
will still cause a reboot afterwards.
.PP
If the machine is halted an email is sent to notify a human that
the machine is going down. Starting with version 4.4 
.B watchdog 
will also notify the human in charge if the machine is rebooted.
.SH "SOFT REBOOT"
A soft reboot (i.e. controlled shutdown and reboot) is initiated for every
error that is found. Since there might be no more processes available,
watchdog does it all by himself. That means:
.IP 1. 4
Kill all processes with SIGTERM.
.IP 2. 4
After a short pause kill all remaining processes with SIGKILL.
.IP 3. 4
Record a shutdown entry in wtmp.
.IP 4. 4
Save the random seed from 
.IR /dev/urandom . 
If the device is non-existant or
there is no filename for saving this step is skipped.
.IP 5. 4
Turn off accounting.
.IP 6. 4
Turn off quota and swap.
.IP 7. 4
Unmount all partitions except the root partition.
.IP 8. 4
Remount the root partition read-only.
.IP 9. 4
Shut down all network interfaces.
.IP 10. 4
Finally reboot.
.SH "CHECK BINARY"
If the return code of the check binary is not zero 
.B watchdog 
will assume an
error and reboot the system. Be careful with this if you are using the
real-time properties of watchdog since 
.B watchdog 
will wait for the return of
this binary before proceeding. An exit code smaller than 245 is interpreted as an
system error code (see 
.I errno.h 
for details). Values of 245 or larger than are special to
.BR watchdog :
.TP
255 (based on \-1 as unsigned 8\-bit number) 
Reboot the system. This is not exactly an error message but a command to
.BR watchdog . 
If the return code is this the 
.B watchdog 
will not try to run a shutdown
script instead.
.TP
254 
Reset the system. This is not exactly an error message but a command to
.BR watchdog . 
If the return code is this the 
.B watchdog will simply refuse to write the
kernel device again.
.TP
253 
Maximum load average exceeded.
.TP
252 
The temperature inside is too high.
.TP
251 
.I /proc/loadavg 
contains no (or not enough) data.
.TP
250 
The given file was not changed in the given interval.
.TP
249 
.I /proc/meminfo 
contains invalid data.
.TP
248
Child process was killed by a signal.
.TP
247
Child process did not return in time.
.TP
246 
Free for personal watchdog-specific use (was \-10 as an unsigned 8\-bit 
number).
.TP
245 
Reserved for an unknown result, for example a slow background test that is 
still running so neither a success nor an error.
.SH "REPAIR BINARY"
The repair binary is started with one parameter: the error number that
caused 
.B watchdog 
to initiate the boot process. After trying to repair the
system the binary should exit with 0 if the system was successfully repaired
and thus there is no need to boot anymore. A return value not equal 0 tells
.B watchdog 
to reboot. The return code of the repair binary should be the error
number of the error causing 
.B watchdog 
to reboot. Be careful with this if you
are using the real-time properties since 
.B watchdog 
will wait for
the return of this binary before proceeding.
.SH "TEST DIRECTORY"
Executables placed in the test directory are discovered by watchdog on 
startup and are automatically executed.  They are bounded time-wise by

ChangeLog
Fri Feb 23 13:26:14 MET 1996

	- Watchdog 1.0 released.

Tue Mar  5 09:11:34 MET 1996

	- Fixed two bugs in watchdog.c reported (and patched) by 
	  Johannes Kroeger <fiction!squirrel.owl.de!jkroeger@golden-gate.owl.de>:
	  Watchdog now tries to open /dev/watchdog before tugging its
	  process ID away.

	- Released version 1.1

Thu May 23 14:20:59 MET DST 1996

	- Added process table watch routine.

	- Process filemode before sleeping.

	- Released version 1.2

Thu May 30 13:00:44 MET DST 1996

	- Watchdog now correctly shuts down the system if the
	  process table is full
 
Mon Jul 29 13:51:42 MET DST 1996

	- Finally watchdog is able to shut down the system even if
	  the process table remains full after killing all processes.

	- Released version 2.0 (this was a major overhaul)

Fri Jan 31 13:39:43 MET 1997

	- Disconnect deamon from console.

	- Close file descriptors 0, 1, 2 when compiled with USE_SYSLOG.

	- Cleaned up source code.

	- Write the softdog device between each check, since the increasing
	  number of checks might cause problems otherise.

	- Test the system load average and reboot if it is too high.

Thu Apr  3 11:14:07 MET DST 1997

	- Fixed that nasty bug that caused the root partion to stay mounted
	  during a soft reboot.

	- Fixed another bug that caused watchdog to kill itself during
	  reboot before all the work was done.

Wed Jun  4 20:16:23 CEST 1997

	- Finally soft reboot works as advertised.

	- Tested NFS access in filemode.

Fri Jun  6 22:55:16 CEST 1997

	- Added ping mode: soft reboot is initiated if a given
	  network/host does not answer.

Wen Jun 11 09:25:36 CEST 1997

	- Added file table check: reboot if file table overflow occurs.

Mon Jun 16 12:10:55 CEST 1997

	- Upgraded to mount 2.6g.
	- Compile and link with glibc.
	- Implemented call to user provides test binary.

Tue Jun 17 13:35:15 CEST 1997

	- Added examples for test scripts.

Wed Jun 18 15:51:10 CEST 1997

	- Default sleep interval now configurable via Makefile.

Fri Jun 20 10:51:56 CEST 1997

	- Added call to repair script.

Tue Jun 24 15:31:03 CEST 1997

	- Repair script now calls logger to create a syslog entry.

Thu Jun 26 21:40:37 CEST 1997

	- Fixed call of repair script.
	- Let watchdog sleep five seconds before closelog() to make sure
          everything is written to the logs.

Mon Jun 30 13:48:08 CEST 1997

	- Make sure files are only opened once.
	- Work even if there is no /dev/watchdog.

Mon Jun 30 21:07:39 CEST 1997

	- Implemented temperature test.
	- Added double fork() to inherit rights from init. Now watchdog will
          unmount all partitions correctly regardless what your current
          working directory is when you start it. My thanks go to Peter Tobias
          <tobias@et-inf.fho-emden.de> for pointing me into the right direction.

Wed Jul  2 11:00:49 CEST 1997

	- Halt the system if temperature is too high, but send mail to the
	  sysadmin prior doing that.

Thu Jul  3 16:22:47 CEST 1997

	- Include fully quantified hostname in Subject: line for mail.

Fri Jul 11 10:38:39 CEST 1997

	- Released version 3.0.

Fri Jul 11 10:49:31 CEST 1997

	- Fixed small typo in Makefile.

Wed Sep  3 12:33:32 CEST 1997

	- Corrected temperature reading code.
	- Include patch to mount.c for sparcs.
	- Write warning to log file when temperature start climbing to high.

Sun Oct 19 11:22:26 CEST 1997

	- No more warnings about unused variables if compiled without
          USE_SYSLOG.
	- Some more glibc adjustments for glibc 2.0.5.

Fri Oct 24 11:40:07 CEST 1997

	- Released version 3.1.

Fri Nov 28 11:10:26 CET 1997

	- Changed handling of return value of repair script to take care of
          negative return value.
	- Stop init before going down.

Fri Nov 28 11:26:50 CET 1997

	- Released version 3.2.

Tue Dec  2 10:02:43 CET 1997

	- Stupid me! Somehow I managed to release my test version without
          removing the call to halt the system regardless of the temperature
          that I used for my last test. I'm sorry. Hope it didn't cause too
          much problems.
	- Released version 3.3.

Tue Dez 30 11:35:03 CET 1997

	- Reorganized watchdog source files. 
	- Implemented no-action option.

Fri Jan  2 16:30:56 CET 1998

	- Made watchdog a real-time application.
	- To enable this I had to change the process-table check code so no
	  hang can occur there.

Thu Sep 10 17:22:32 CEST 1998

	- Updated mount code to version 2.8 something. Thanks to the one who
	  send me the patches. Unfortunately I lost his email-address.

Fri Sep 11 16:22:28 CEST 1998

	- Updated test binary stuff to make sure it won't reboot the system
	  if binary runs for too long.

Tue Sep 15 13:28:35 CEST 1998

	- Allow multiple filenames resp. ip-addresses.

Fri Sep 18 16:10:56 CEST 1998

	- Move all options with arguments to watchdog.conf and added long
	  options for the remaining ones.
	- Added option -c to add config file name.

Mon Sep 21 13:16:15 CEST 1998

	- Output file modification time in verbose mode.
	- Added check for constant changing of a file.

Mon Sep 21 15:23:54 CEST 1998

	- Updated man pages.

Thu Oct 15 18:21:57 CEST 1998

	- Released version 4.0.

Fri Oct 16 14:22:15 CEST 1998

	- Received patches for RedHat. Argh, should have waited with
	  releasing 4.0.
	- Some more modifications to repair.sh.

Wed Okt 21 13:33:21 CEST 1998

	- Moved configuration to config.h.

Tue Okt 27 20:49:23 MEZ 1998

	- Released 4.1.

Mon Nov 23 09:33:20 MEZ 1998

	- Changed union name in extern.h from options to wdog_options to be
	  able to compile on libc5 systems.

Tue Nov 24 14:32:41 MEZ 1998

	- Added patches by Sascha Schumann <sas@schell.de> for libc5
	  compatibility.

Thu Jan 14 13:43:53 CET 1999

	- Fixed the bug in read_config that caused the max-load-15 line to
	  be read as max-load-1 = 5.

Thu Jan 14 14:17:35 CET 1999

	- And fixed another silly bug that caused ping mode to not work
	  anymore.

Wed Jan 20 21:03:30 CET 1999


src/watchdog.c
/*************************************************************/
/* Original version was an example in the kernel source tree */
/*                                                           */
/* Most of the rest was written by me, Michael Meskes        */
/* meskes@debian.org                                         */
/*                                                           */
/*************************************************************/
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <signal.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <string.h>
#include <time.h>
#include <libgen.h>
#include <unistd.h>
#include <sys/param.h>		/* For EXEC_PAGESIZE */

#include "watch_err.h"
#include "extern.h"
#include "read-conf.h"		/* For add_list() */
#include "timefunc.h"

static int no_act = FALSE;

static void usage(char *progname)
{
	fprintf(stderr, "%s version %d.%d, usage:\n", progname, MAJOR_VERSION, MINOR_VERSION);
	fprintf(stderr, "%s [options]\n", progname);
	fprintf(stderr, "options:\n");
	fprintf(stderr, "  -c | --config-file <file>  specify location of config file\n");
	fprintf(stderr, "  -f | --force               don't sanity-check config or use PID file\n");
	fprintf(stderr, "  -F | --foreground          run in foreground\n");
	fprintf(stderr, "  -X | --loop-exit <number>  run a fixed number of loops then exit\n");
	fprintf(stderr, "  -q | --no-action           do not reboot or halt\n");
	fprintf(stderr, "  -b | --softboot            soft-boot on error\n");
	fprintf(stderr, "  -s | --sync                sync filesystem\n");
	fprintf(stderr, "  -v | --verbose             verbose messages\n");
	exit(1);
}

/* Try to sync */
static int sync_system(int sync_it)
{
	if (sync_it) {
		sync();
		sync();
	}
	return (0);
}

/* execute repair binary */
static int repair(char *rbinary, int result, char *name, int version)
{
	char *arg[6];
	char parm[22];
	int ret;

	sprintf(parm, "%d", result);

	if (version == 0) {
		arg[0] = rbinary;	/* Use common repair binary with V0 test scripts, etc. */
		arg[1] = rbinary;
		arg[2] = parm;
		arg[3] = name;	/* May be null, not a problem here. */
		arg[4] = NULL;
		arg[5] = NULL;
	} else {
		arg[0] = name;	/* With V1 the test binary is also the repair binary. */
		arg[1] = name;
		arg[2] = "repair";
		arg[3] = parm;
		arg[4] = name;
		arg[5] = NULL;
	}

	/* no binary given, we have to reboot */
	if (arg[0] == NULL)
		return (result);

	ret = run_func_as_child(repair_timeout, exec_as_func, FLAG_REOPEN_STD_REPAIR, arg);

	/* check result */
	if (ret != 0) {
		log_message(LOG_ERR, "repair binary %s returned %d = '%s'", rbinary, ret, wd_strerror(ret));
	}

	return (ret);
}

static void wd_action(int result, char *rbinary, struct list *act)
{
	int version = 0;
	char *name = NULL;
	int timeout = TRUE;

	/* If we have info about the version, use this to decide what to call
	 * in order to repair the problem. Default is we use the global repair
	 * call, but for V1 we use the same program with a different argument.
	 */
	if (act != NULL) {
		name = act->name;
		version = act->version;
	}

	if (version == 1) {
		rbinary = name;
	}

	/* Decide on repair or return based on error code. */
	switch (result) {
	case ENOERR:
		/* No error, reset any time-out. */
		if (act != NULL) {
			act->last_time = 0;
			act->repair_count = 0;
		}
	case EDONTKNOW:
		/* Don't know, keep on working */
		return;

	case EREBOOT:
	case ERESET:
	case ETOOHOT:
	case EMAXLOAD:	/* System too busy? */
	case EMFILE:	/* "Too many open files" */
	case ENFILE:	/* "Too many open files in system" */
	case ENOMEM:	/* "Not enough space" */
		/* These are not repairable. */
		break;

	default:
		/* error that might be repairable */
		if (act != NULL && retry_timeout > 0) {
			/* timer possible and used to allow re-try */
			time_t now = time_mono(NULL);
			timeout = FALSE;

			if (act->last_time == 0) {
				/* First offence, record time. */
				act->last_time = now;
			} else {
				/* timer running */
				int tused = (int)(now - act->last_time);

				if (tused > retry_timeout) {
					log_message(LOG_WARNING, "Retry timed-out at %d seconds for %s", tused,
						act->name);
					timeout = TRUE;
				} else {
					if (verbose)
						log_message(LOG_DEBUG, "Retry at %d seconds for %s", tused, act->name);
				}
			}
		}

		if (timeout) {
			int try_repair = TRUE;
			/* check for too many failed repair attempts */
			if (act != NULL && repair_max > 0) {
				if (++act->repair_count > repair_max) {
					try_repair = FALSE;
					log_message(LOG_WARNING, "Repair count exceeded (%d for %s)",
						act->repair_count, act->name);
				} else {
					/* going to repair, reset re-try timer so same period for next try */
					act->last_time = 0;
					if (verbose) {
						log_message(LOG_DEBUG, "Repair attempt %d for %s",
							act->repair_count, act->name);
					}
				}
			}

			if (try_repair) {
				result = repair(rbinary, result, name, version);
			}
		} else {
			result = ENOERR;
		}
		break;
	}

	/* if still error, consider reboot */
	if (result != ENOERR) {
		/* if no-action flag set, do nothing */
		if (no_act) {
			if (verbose) {
				log_message(LOG_DEBUG, "Shutdown blocked by --no-action (error %d = '%s')",
					result, wd_strerror(result));
			}
		} else {
			do_shutdown(result);
		}
	}
}

static void do_check(int res, char *rbinary, struct list *act)
{
	wd_action(res, rbinary, act);
	wd_action(keep_alive(), rbinary, NULL);
}

static void old_option(int c, char *configfile)
{
	fprintf(stderr, "Option -%c is no longer valid, please specify it in %s.\n", c, configfile);
}

static void print_info(int sync_it, int force)
{
	struct list *act;

	log_message(LOG_INFO, "int=%ds realtime=%s sync=%s load=%d,%d,%d",
		    tint,
		    realtime ? "yes" : "no",
		    sync_it ? "yes" : "no",
		    maxload1, maxload5, maxload15);

	if (minpages == 0 && minalloc == 0)
		log_message(LOG_INFO, "memory not checked");
	else
		log_message(LOG_INFO, "memory: minimum pages = %d free, %d allocatable (%d byte pages)",
			minpages, minalloc, EXEC_PAGESIZE);

	if (target_list == NULL)
		log_message(LOG_INFO, "ping: no machine to check");
	else
		for (act = target_list; act != NULL; act = act->next)
			log_message(LOG_INFO, "ping: %s", act->name);

	if (file_list == NULL)
		log_message(LOG_INFO, "file: no file to check");
	else
		for (act = file_list; act != NULL; act = act->next)
			log_message(LOG_INFO, "file: %s:%d", act->name, act->parameter.file.mtime);

	if (pidfile_list == NULL)
		log_message(LOG_INFO, "pidfile: no server process to check");
	else
		for (act = pidfile_list; act != NULL; act = act->next)
			log_message(LOG_INFO, "pidfile: %s", act->name);

	if (iface_list == NULL)
		log_message(LOG_INFO, "interface: no interface to check");
	else
		for (act = iface_list; act != NULL; act = act->next)
			log_message(LOG_INFO, "interface: %s", act->name);

	if (temp_list == NULL)
		log_message(LOG_INFO, "temperature: no sensors to check");
	else {
		log_message(LOG_INFO, "temperature: maximum = %d", maxtemp);
		for (act = temp_list; act != NULL; act = act->next)
			log_message(LOG_INFO, "temperature: %s", act->name);
	}

	if (tr_bin_list == NULL)
		log_message(LOG_INFO, "no test binary files");
	else {
		log_message(LOG_INFO, "test binary time-out = %d", test_timeout);
		for (act = tr_bin_list; act != NULL; act = act->next)
			log_message(LOG_INFO, "%s: %s",
				act->version == 0 ? "test binary V0" : "test/repair V1",
				act->name);
	}

	if (repair_bin == NULL)
		log_message(LOG_INFO, "no repair binary files");
	else {
		log_message(LOG_INFO, "repair binary: time-out = %d", repair_timeout);
		log_message(LOG_INFO, "repair binary: program = %s", repair_bin);
	}

	log_message(LOG_INFO, "error retry time-out = %d seconds", retry_timeout);

	if (repair_max > 0) {
		log_message(LOG_INFO, "repair attempts = %d", repair_max);
	} else {
		log_message(LOG_INFO, "repair attempts = unlimited");
	}

	log_message(LOG_INFO, "alive=%s heartbeat=%s to=%s no_act=%s force=%s",
		    (devname == NULL) ? "[none]" : devname,
		    (heartbeat == NULL) ? "[none]" : heartbeat,
		    (admin == NULL) ? "[none]" : admin,
		    (no_act == TRUE) ? "yes" : "no",
		    (force == TRUE) ? "yes" : "no");
}

/*
 * Check some of the configured parameters for sensibility. If any look bad, exit with
 * the message about using the --force option to skip these checks.
 */

static void check_parameters(void)
{
	int err = 0;
	int min_timeout = 1;
	int max_timeout = 15 * dev_timeout;
	int temp_limit = 55;	/* Lower limit on temperature in Celsius. */

	if (max_timeout < 120) {
		max_timeout = 120;
	}

	/* check polling interval against the hardware timer value - VERY important. */
	if (dev_timeout - tint < 2) {
		log_message(LOG_ERR,
			    "Error: This interval length (%d) might reboot the system while the process sleeps!",
			    tint);
		err = 1;
	}

	/* check the 3 load average limits. */
	if (maxload1 > 0 && maxload1 < MINLOAD) {
		log_message(LOG_ERR,
			    "Error: Using this maximal 1-minute load average (%d) might reboot the system too often!",
			    maxload1);
		err = 1;
	}

	if (maxload5 > 0 && maxload5 < MINLOAD) {
		log_message(LOG_ERR,
			    "Error: Using this maximal 5-minute load average (%d) might reboot the system too often!",
			    maxload5);
		err = 1;
	}

	if (maxload15 > 0 && maxload15 < MINLOAD) {
		log_message(LOG_ERR,
			    "Error: Using this maximal 15-minute load average (%d) might reboot the system too often!",
			    maxload15);
		err = 1;
	}

	/* check the other time-outs, these can be independent of the hardware timer, but apply some sense! */
	if (repair_timeout > max_timeout || repair_timeout < min_timeout) {
		log_message(LOG_ERR,
			    "Error: This repair time-out (%d) looks out of a sensible range (%d..%d)!",
			    repair_timeout, min_timeout, max_timeout);
		err = 1;
	}

	if (test_timeout > max_timeout || test_timeout < min_timeout) {
		log_message(LOG_ERR,
			    "Error: This test time-out (%d) looks out of a sensible range (%d..%d)!",
			    test_timeout, min_timeout, max_timeout);
		err = 1;
	}

	/* unlike the other time-outs, setting retry time-out to zero is not totally unreasonable. */
	min_timeout = 0;
	if (retry_timeout > max_timeout || retry_timeout < min_timeout) {
		log_message(LOG_ERR,
			    "Error: This retry time-out (%d) looks out of a sensible range (%d..%d)!",
			    retry_timeout, min_timeout, max_timeout);
		err = 1;
	}

	if (temp_list != NULL && maxtemp < temp_limit) {
		log_message(LOG_ERR,
			    "Error: Max temperature of %d is too low to be sensible (test = %dC)",
			    maxtemp, temp_limit);
		err = 1;
	}

	if (err) {
		fatal_error(EX_USAGE, "To force parameter(s) use the --force command line option.");
	}
}

int main(int argc, char *const argv[])
{
	int c, foreground = FALSE, force = FALSE, sync_it = FALSE;
	char *configfile = CONFIG_FILENAME;
	struct list *act;
	char *progname;
	char *opts = "d:i:n:Ffsvbql:p:t:c:r:m:a:X:";
	struct option long_options[] = {
		{"config-file", required_argument, NULL, 'c'},
		{"foreground", no_argument, NULL, 'F'},
		{"force", no_argument, NULL, 'f'},
		{"sync", no_argument, NULL, 's'},
		{"no-action", no_argument, NULL, 'q'},
		{"verbose", no_argument, NULL, 'v'},
		{"softboot", no_argument, NULL, 'b'},
		{"loop-exit", required_argument, NULL, 'X'},
		{NULL, 0, NULL, 0}
	};
	long count = 0L;
	long count_max = 0L;
	unsigned long swait, twait;
	int softboot = FALSE;
	struct list *memtimer = NULL;
	struct list *loadtimer = NULL;

	progname = basename(argv[0]);
	open_logging(progname, MSG_TO_STDERR | MSG_TO_SYSLOG);

	/* check the options */
	/* there aren't that many any more */
	while ((c = getopt_long(argc, argv, opts, long_options, NULL)) != EOF) {
		switch (c) {
		case 'n':
		case 'p':
		case 'a':
		case 'r':
		case 'd':
		case 't':
		case 'l':
		case 'm':
		case 'i':
			old_option(c, configfile);
			usage(progname);
			break;
		case 'c':
			configfile = optarg;
			break;
		case 'F':
			foreground = TRUE;
			break;
		case 'f':
			force = TRUE;
			break;
		case 's':
			sync_it = TRUE;
			break;
		case 'b':
			softboot = TRUE;
			break;
		case 'q':
			no_act = TRUE;
			break;
		case 'v':
			verbose++;
			break;
		case 'X':
			count_max = atol(optarg);
			log_message(LOG_WARNING, "NOTE: Using --loop-exit so daemon will exit after %ld time intervals",
				    count_max);
			break;
		default:
			usage(progname);
		}
	}

	add_list(&memtimer, "<free-memory>", 0);
	add_list(&loadtimer, "<load-average>", 0);

	read_config(configfile);

	if (softboot) {
		/* Result of zeroing time-out is immediate action to shut down on errors, rather like old softboot behaviour. */
		retry_timeout = 0;
	}

	if (!force) {
		check_parameters();
	}

	/* make sure we get our own log directory */
	if (mkdir(logdir, 0750) && errno != EEXIST) {
		fatal_error(EX_SYSERR, "Cannot create directory %s (%s)", logdir, strerror(errno));
	}

	/* set up pinging if in ping mode */
	if (target_list != NULL) {
		open_netcheck(target_list);
	}

	if (!foreground) {
		/* allocate some memory to store a filename, this is needed later on even
		 * if the system runs out of memory */
		set_reopen_dir(logdir);

		if (wd_daemon(0, 0)) {
			fatal_error(EX_SYSERR, "failed to daemonize (%s)", strerror(errno));
		}
		open_logging(NULL, MSG_TO_SYSLOG);	/* Close terminal output, keep syslog open. */
	}

	/* tuck my process id away */
	if (!force && write_pid_file(PIDFILE)) {
		fatal_error(EX_USAGE, "unable to gain lock via PID file");
	}

	/* Log the starting message */
	log_message(LOG_NOTICE, "starting daemon (%d.%d):", MAJOR_VERSION, MINOR_VERSION);
	print_info(sync_it, force);

	/* open the device */
	if (no_act == FALSE) {
		open_watchdog(devname, dev_timeout);
	}

	open_tempcheck(temp_list);

	open_heartbeat();

	open_loadcheck();

	open_memcheck();

	/* set signal term to set our run flag to 0 so that */
	/* we make sure watchdog device is closed when receiving SIGTERM */
	signal(SIGTERM, sigterm_handler);

	lock_our_memory(realtime, schedprio, daemon_pid);

	/* Short wait (50ms OK?) in case test binaries return quickly, then
	 * remaining 'twait' should make watchdog sleep 'tint' seconds total.
	 */
	swait = 50000;
	twait = (tint * 1000000) - swait;

	/* main loop: update after <tint> seconds */
	while (_running) {
		wd_action(keep_alive(), repair_bin, NULL);

		/* sync system if we have to */
		do_check(sync_system(sync_it), repair_bin, NULL);

		/* check file table */
		do_check(check_file_table(), repair_bin, NULL);

		/* check load average */
		do_check(check_load(), repair_bin, loadtimer);

		/* check free memory */
		do_check(check_memory(), repair_bin, memtimer);

		/* check allocatable memory */
		do_check(check_allocatable(), repair_bin, memtimer);

		/* check temperature */
		for (act = temp_list; act != NULL; act = act->next)
			do_check(check_temp(act), repair_bin, act);

		/* in filemode stat file */
		for (act = file_list; act != NULL; act = act->next)
			do_check(check_file_stat_safe(act), repair_bin, act);

		/* in pidmode kill -0 processes */
		for (act = pidfile_list; act != NULL; act = act->next)
			do_check(check_pidfile(act), repair_bin, act);

		/* in network mode check the given devices for input */
		for (act = iface_list; act != NULL; act = act->next)
			do_check(check_iface(act), repair_bin, act);

		/* in ping mode ping the ip address */
		for (act = target_list; act != NULL; act = act->next)
			do_check(check_net(act->name,
					   act->parameter.net.sock_fp,
					   act->parameter.net.to,
					   act->parameter.net.packet, tint, pingcount), repair_bin, act);

		/* test, or test/repair binaries in the watchdog.d directory */
		for (act = tr_bin_list; act != NULL; act = act->next)
			do_check(check_bin(act->name, test_timeout, act->version), repair_bin, act);

		/* in case test binaries return quickly */
		usleep(swait);
		check_bin(NULL, test_timeout, 0);

		/* finally sleep for a full cycle */
		/* we have just triggered the device with the last check */
		usleep(twait);

		count++;

		/* do verbose logging */
		if (verbose && logtick && (--ticker == 0)) {
			ticker = logtick;
			log_message(LOG_DEBUG, "still alive after %ld interval(s)", count);
		}

		if (count_max > 0 && count >= count_max) {
			log_message(LOG_WARNING, "loop exit on interval counter reached");
			_running = 0;
		}
	}

	free_list(&loadtimer);
	free_list(&memtimer);

	terminate(EXIT_SUCCESS);
	/* not reached */
	return (EXIT_SUCCESS);
}
